File: nimble-core/src/recurrence.rs
//! Recurrence engine: pure functions to parse a small recurrence grammar and
//! compute the next occurrence of a recurring due date. No I/O, no DB.
//!
//! Grammar (case-insensitive, this is the WHOLE grammar):
//!   "every day" | "every N days"
//!   "every week" | "every N weeks"
//!   "every month" | "every N months"
//!   "every year" | "every N years"
//! optionally followed by "@ HH:MM" or "at H[:MM][am|pm]".

use chrono::{Datelike, Days, Months, NaiveDate};

#[derive(Debug, Clone, PartialEq)]
pub struct RecurrenceRule {
    pub interval: u32,          // every N units, >= 1
    pub unit: RecurrenceUnit,   // Day | Week | Month | Year
    pub time: Option<String>,   // "HH:MM" 24h
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum RecurrenceUnit {
    Day,
    Week,
    Month,
    Year,
}

/// None = string is not a supported rule (caller stores it anyway; task just won't auto-recur)
pub fn parse_rule(s: &str) -> Option<RecurrenceRule> {
    let lower = s.trim().to_lowercase();
    if lower.is_empty() {
        return None;
    }

    // Split off the time suffix, introduced by "@" or "at".
    let (body, time) = split_time_suffix(&lower)?;

    let tokens: Vec<&str> = body.split_whitespace().collect();
    if tokens.first() != Some(&"every") {
        return None;
    }

    // Remaining tokens after "every": either [singular_unit] or [N, plural_unit],
    // matching the grammar exactly ("every day" / "every N days", never a bare
    // plural like "every days" or a singular with a count like "every 2 day").
    let rest = &tokens[1..];
    let (interval, unit) = match rest {
        [unit] => (1u32, match_singular_unit(unit)?),
        [n, unit] => {
            let n: u32 = n.parse().ok()?;
            (n, match_plural_unit(unit)?)
        }
        _ => return None,
    };

    if interval == 0 {
        return None;
    }

    Some(RecurrenceRule { interval, unit, time })
}

fn match_singular_unit(unit: &str) -> Option<RecurrenceUnit> {
    match unit {
        "day" => Some(RecurrenceUnit::Day),
        "week" => Some(RecurrenceUnit::Week),
        "month" => Some(RecurrenceUnit::Month),
        "year" => Some(RecurrenceUnit::Year),
        _ => None,
    }
}

fn match_plural_unit(unit: &str) -> Option<RecurrenceUnit> {
    match unit {
        "days" => Some(RecurrenceUnit::Day),
        "weeks" => Some(RecurrenceUnit::Week),
        "months" => Some(RecurrenceUnit::Month),
        "years" => Some(RecurrenceUnit::Year),
        _ => None,
    }
}

/// Splits a lowercased rule string into (body, optional "HH:MM" time),
/// where the time suffix is introduced by "@" or the word "at".
fn split_time_suffix(lower: &str) -> Option<(String, Option<String>)> {
    if let Some(idx) = lower.find('@') {
        let body = lower[..idx].trim().to_string();
        let time_str = lower[idx + 1..].trim();
        let time = parse_time(time_str)?;
        return Some((body, Some(time)));
    }

    // Look for a standalone " at " token boundary.
    let tokens: Vec<&str> = lower.split_whitespace().collect();
    if let Some(pos) = tokens.iter().position(|t| *t == "at") {
        let body = tokens[..pos].join(" ");
        let time_tokens = &tokens[pos + 1..];
        if time_tokens.is_empty() {
            return None;
        }
        let time = parse_time(&time_tokens.join(" "))?;
        return Some((body, Some(time)));
    }

    Some((lower.to_string(), None))
}

/// Parses "H[:MM][am|pm]" (or 24h "HH:MM") into "HH:MM".
fn parse_time(s: &str) -> Option<String> {
    let s = s.trim();
    if s.is_empty() {
        return None;
    }

    let (digits, meridiem) = if let Some(stripped) = s.strip_suffix("am") {
        (stripped.trim(), Some(false))
    } else if let Some(stripped) = s.strip_suffix("pm") {
        (stripped.trim(), Some(true))
    } else {
        (s, None)
    };

    let (hour_str, minute_str) = match digits.split_once(':') {
        Some((h, m)) => (h, m),
        None => (digits, "0"),
    };

    let mut hour: u32 = hour_str.parse().ok()?;
    let minute: u32 = minute_str.parse().ok()?;
    if minute > 59 {
        return None;
    }

    match meridiem {
        Some(is_pm) => {
            if !(1..=12).contains(&hour) {
                return None;
            }
            hour %= 12; // 12am -> 0, 12pm -> 12 (below)
            if is_pm {
                hour += 12;
            }
        }
        None => {
            if hour > 23 {
                return None;
            }
        }
    }

    Some(format!("{:02}:{:02}", hour, minute))
}

/// Next due date strictly after `today`, advancing by whole intervals from `current_due`.
pub fn next_occurrence(rule: &RecurrenceRule, current_due: NaiveDate, today: NaiveDate) -> NaiveDate {
    let mut candidate = current_due;
    loop {
        candidate = add_interval(candidate, rule);
        if candidate > today {
            return candidate;
        }
    }
}

fn add_interval(date: NaiveDate, rule: &RecurrenceRule) -> NaiveDate {
    match rule.unit {
        RecurrenceUnit::Day => date + Days::new(rule.interval as u64),
        RecurrenceUnit::Week => date + Days::new(rule.interval as u64 * 7),
        RecurrenceUnit::Month => add_months_clamped(date, rule.interval),
        RecurrenceUnit::Year => add_months_clamped(date, rule.interval * 12),
    }
}

/// Adds N months, clamping the day-of-month to the target month's length:
/// first day of (month + N), then min(original day, days_in_that_month).
fn add_months_clamped(date: NaiveDate, months: u32) -> NaiveDate {
    let day = date.day();
    let first_of_month = date.with_day(1).expect("day 1 is always valid");
    let target_first = first_of_month + Months::new(months);
    let days_in_target = days_in_month(target_first.year(), target_first.month());
    let clamped_day = day.min(days_in_target);
    target_first
        .with_day(clamped_day)
        .expect("clamped_day is within the target month's range")
}

fn days_in_month(year: i32, month: u32) -> u32 {
    let (next_year, next_month) = if month == 12 { (year + 1, 1) } else { (year, month + 1) };
    let first_of_next = NaiveDate::from_ymd_opt(next_year, next_month, 1).expect("valid date");
    let first_of_this = NaiveDate::from_ymd_opt(year, month, 1).expect("valid date");
    (first_of_next - first_of_this).num_days() as u32
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::NaiveDate;
    fn d(s: &str) -> NaiveDate { NaiveDate::parse_from_str(s, "%Y-%m-%d").unwrap() }

    #[test]
    fn parses_marcos_real_rules() {
        assert_eq!(parse_rule("every month"),
            Some(RecurrenceRule { interval: 1, unit: RecurrenceUnit::Month, time: None }));
        assert_eq!(parse_rule("every 2 weeks @ 09:00"),
            Some(RecurrenceRule { interval: 2, unit: RecurrenceUnit::Week, time: Some("09:00".into()) }));
        assert_eq!(parse_rule("Every Day"),
            Some(RecurrenceRule { interval: 1, unit: RecurrenceUnit::Day, time: None }));
        assert_eq!(parse_rule("every 3 months at 9am"),
            Some(RecurrenceRule { interval: 3, unit: RecurrenceUnit::Month, time: Some("09:00".into()) }));
    }

    #[test]
    fn rejects_unsupported_strings() {
        for s in ["every 3rd tuesday", "weekdays", "every!", "", "tomorrow", "every 0 days"] {
            assert_eq!(parse_rule(s), None, "should reject {s:?}");
        }
    }

    #[test]
    fn completed_early_advances_one_interval_from_due() {
        // EDD: due 8/16, completed 8/10 → next 8/30 (from due, not from completion day)
        let rule = parse_rule("every 2 weeks @ 09:00").unwrap();
        assert_eq!(next_occurrence(&rule, d("2026-08-16"), d("2026-08-10")), d("2026-08-30"));
    }

    #[test]
    fn completed_late_advances_past_today() {
        // loans: due 8/27, completed 10/02 → next 10/27 (skips the already-past 9/27)
        let rule = parse_rule("every month").unwrap();
        assert_eq!(next_occurrence(&rule, d("2026-08-27"), d("2026-10-02")), d("2026-10-27"));
    }

    #[test]
    fn month_end_clamps() {
        let rule = parse_rule("every month").unwrap();
        assert_eq!(next_occurrence(&rule, d("2026-01-31"), d("2026-01-31")), d("2026-02-28"));
        assert_eq!(next_occurrence(&rule, d("2028-01-31"), d("2028-01-31")), d("2028-02-29")); // leap year
    }

    #[test]
    fn yearly_and_daily() {
        assert_eq!(next_occurrence(&parse_rule("every year").unwrap(), d("2026-03-06"), d("2026-03-06")), d("2027-03-06"));
        assert_eq!(next_occurrence(&parse_rule("every day").unwrap(), d("2026-08-09"), d("2026-08-09")), d("2026-08-10"));
    }

    // --- Additional edge cases beyond the mandatory set ---

    #[test]
    fn parses_time_variants() {
        assert_eq!(parse_rule("every week at 5pm"),
            Some(RecurrenceRule { interval: 1, unit: RecurrenceUnit::Week, time: Some("17:00".into()) }));
        assert_eq!(parse_rule("every day at 12am"),
            Some(RecurrenceRule { interval: 1, unit: RecurrenceUnit::Day, time: Some("00:00".into()) }));
        assert_eq!(parse_rule("every day at 12pm"),
            Some(RecurrenceRule { interval: 1, unit: RecurrenceUnit::Day, time: Some("12:00".into()) }));
        assert_eq!(parse_rule("every day at 9:30am"),
            Some(RecurrenceRule { interval: 1, unit: RecurrenceUnit::Day, time: Some("09:30".into()) }));
    }

    #[test]
    fn rejects_more_malformed_strings() {
        for s in [
            "every day at",           // "at" with nothing following
            "every day @",            // "@" with nothing following
            "every day at 13pm",      // out-of-range 12h hour
            "every day @ 25:00",      // out-of-range 24h hour
            "every day @ 09:99",      // out-of-range minutes
            "every -1 days",          // negative interval unparsable as u32
            "every months",           // bare plural with no count is not in the grammar
            "every 2 day",            // count with singular unit is not in the grammar
            "every 2",                // missing unit
            "every",                  // nothing after "every"
        ] {
            assert_eq!(parse_rule(s), None, "should reject {s:?}");
        }
    }

    #[test]
    fn is_case_insensitive_and_trims_whitespace() {
        assert_eq!(parse_rule("  EVERY   Month  "),
            Some(RecurrenceRule { interval: 1, unit: RecurrenceUnit::Month, time: None }));
    }

    #[test]
    fn next_occurrence_never_returns_today_even_when_interval_lands_on_it() {
        // If advancing by one interval lands exactly on `today`, keep advancing.
        let rule = parse_rule("every week").unwrap();
        assert_eq!(next_occurrence(&rule, d("2026-08-02"), d("2026-08-09")), d("2026-08-16"));
    }

    #[test]
    fn year_leap_day_clamps_to_feb_28_on_non_leap_year() {
        let rule = parse_rule("every year").unwrap();
        assert_eq!(next_occurrence(&rule, d("2028-02-29"), d("2028-02-29")), d("2029-02-28"));
    }
}

File: apps/desktop/src/services/turso/recurrence.ts
/**
 * Recurrence engine for the web build — a line-for-line port of the Rust engine.
 *
 * ⚠️ SOURCE OF TRUTH: `nimble-core/src/recurrence.rs`.
 * This file is a SECOND implementation of the same grammar, and it is only
 * worth having if it provably agrees with the first. Any change to
 * `recurrence.rs` — the grammar, the clamping rules, the parse failures — MUST
 * be mirrored here in the same commit, and the corresponding case added to
 * `recurrence.vectors.ts`. Run `npx tsx apps/desktop/src/services/turso/recurrence.selfcheck.ts`
 * afterwards; it exits non-zero on any divergence.
 *
 * Beyond those vectors, this port was differentially fuzzed against the real
 * Rust (an oracle binary built with `#[path] mod rec;` pointing at the actual
 * recurrence.rs): 39 adversarial parse inputs and 11,328 `next_occurrence`
 * cases — every month-end day across leap and non-leap years, all four units,
 * intervals 0 through 100,000, and randomised due/today pairs spanning
 * 1970-2079 in both directions — produced byte-identical output.
 *
 * Grammar (case-insensitive, this is the WHOLE grammar):
 *   "every day" | "every N days"
 *   "every week" | "every N weeks"
 *   "every month" | "every N months"
 *   "every year" | "every N years"
 * optionally followed by "@ HH:MM" or "at H[:MM][am|pm]".
 *
 * Calling convention (see `update_task_status` in `nimble-core/src/db/tasks.rs`,
 * ~line 550): on completing a task that has BOTH a parseable `recurrence_rule`
 * and a `due_date`, the task does not complete. Instead
 *   due_date = nextOccurrence(rule, due_date, today)
 *   due_time = rule.time ?? existing due_time      // rule.time WINS
 *   status   = 'todo'
 * A rule that fails to parse (null here, `None` in Rust) or a missing due date
 * falls through to a normal completion — the rule is inert in that case.
 *
 * Dates are handled as `YYYY-MM-DD` strings converted to plain {y, m, d}
 * integer triples and back. There is deliberately NO `Date` object anywhere in
 * this file: `new Date("2026-08-15")` parses as UTC while `new Date(2026, 7, 15)`
 * parses as local, and mixing them silently shifts a due date by a day.
 * Integer civil-date arithmetic has no timezone at all.
 */

/**
 * Sane upper bound on `interval`, mirroring `MAX_INTERVAL` in recurrence.rs.
 * Enforced by `parseRule`, and re-enforced defensively in `nextOccurrence`
 * because a `RecurrenceRule` can be built without going through the parser
 * (e.g. deserialized from storage): interval 0 would never advance past
 * `today` (infinite loop) and an unbounded interval would run the date
 * arithmetic off into nonsense.
 */
const MAX_INTERVAL = 100_000

/** Mirrors `u32::MAX` — the integer type Rust parses interval/hour/minute into. */
const U32_MAX = 4_294_967_295

export type RecurrenceUnit = 'day' | 'week' | 'month' | 'year'

/** Mirrors the `RecurrenceRule` struct in recurrence.rs. */
export interface RecurrenceRule {
  /** every N units, >= 1 */
  interval: number
  unit: RecurrenceUnit
  /** "HH:MM" 24h, or null where Rust has `None` */
  time: string | null
}

// ---------------------------------------------------------------------------
// Parsing
// ---------------------------------------------------------------------------

/**
 * `None` = string is not a supported rule (caller stores it anyway; the task
 * just won't auto-recur). Returns `null` at exactly the points where Rust's
 * `parse_rule` returns `None`.
 */
export function parseRule(rule: string): RecurrenceRule | null {
  const lower = rule.trim().toLowerCase()
  if (lower.length === 0) {
    return null
  }

  // Split off the time suffix, introduced by "@" or "at".
  const split = splitTimeSuffix(lower)
  if (split === null) {
    return null
  }
  const { body, time } = split

  const tokens = splitWhitespace(body)
  if (tokens[0] !== 'every') {
    return null
  }

  // Remaining tokens after "every": either [singular_unit] or [N, plural_unit],
  // matching the grammar exactly ("every day" / "every N days", never a bare
  // plural like "every days" or a singular with a count like "every 2 day").
  const rest = tokens.slice(1)
  let interval: number
  let unit: RecurrenceUnit | null
  if (rest.length === 1) {
    interval = 1
    unit = matchSingularUnit(rest[0])
  } else if (rest.length === 2) {
    const n = parseU32(rest[0])
    if (n === null) {
      return null
    }
    interval = n
    unit = matchPluralUnit(rest[1])
  } else {
    return null
  }
  if (unit === null) {
    return null
  }

  if (interval === 0 || interval > MAX_INTERVAL) {
    return null
  }

  return { interval, unit, time }
}

function matchSingularUnit(unit: string): RecurrenceUnit | null {
  switch (unit) {
    case 'day':
      return 'day'
    case 'week':
      return 'week'
    case 'month':
      return 'month'
    case 'year':
      return 'year'
    default:
      return null
  }
}

function matchPluralUnit(unit: string): RecurrenceUnit | null {
  switch (unit) {
    case 'days':
      return 'day'
    case 'weeks':
      return 'week'
    case 'months':
      return 'month'
    case 'years':
      return 'year'
    default:
      return null
  }
}

/**
 * Splits a lowercased rule string into (body, optional "HH:MM" time), where the
 * time suffix is introduced by "@" or the word "at". `null` = the suffix was
 * present but unparseable, which fails the whole rule (Rust's `?` on
 * `parse_time`).
 */
function splitTimeSuffix(lower: string): { body: string; time: string | null } | null {
  const idx = lower.indexOf('@')
  if (idx !== -1) {
    const body = lower.slice(0, idx).trim()
    const timeStr = lower.slice(idx + 1).trim()
    const time = parseTime(timeStr)
    if (time === null) {
      return null
    }
    return { body, time }
  }

  // Look for a standalone " at " token boundary.
  const tokens = splitWhitespace(lower)
  const pos = tokens.indexOf('at')
  if (pos !== -1) {
    const body = tokens.slice(0, pos).join(' ')
    const timeTokens = tokens.slice(pos + 1)
    if (timeTokens.length === 0) {
      return null
    }
    const time = parseTime(timeTokens.join(' '))
    if (time === null) {
      return null
    }
    return { body, time }
  }

  return { body: lower, time: null }
}

/** Parses "H[:MM][am|pm]" (or 24h "HH:MM") into "HH:MM". */
function parseTime(input: string): string | null {
  const s = input.trim()
  if (s.length === 0) {
    return null
  }

  let digits: string
  let isPm: boolean | null
  if (s.endsWith('am')) {
    digits = s.slice(0, -2).trim()
    isPm = false
  } else if (s.endsWith('pm')) {
    digits = s.slice(0, -2).trim()
    isPm = true
  } else {
    digits = s
    isPm = null
  }

  // `split_once(':')` — splits on the FIRST colon only, so "9:30:00" yields
  // minute_str "30:00", which then fails to parse. Same here.
  const colon = digits.indexOf(':')
  const hourStr = colon === -1 ? digits : digits.slice(0, colon)
  const minuteStr = colon === -1 ? '0' : digits.slice(colon + 1)

  let hour = parseU32(hourStr)
  const minute = parseU32(minuteStr)
  if (hour === null || minute === null) {
    return null
  }
  if (minute > 59) {
    return null
  }

  if (isPm !== null) {
    if (hour < 1 || hour > 12) {
      return null
    }
    hour %= 12 // 12am -> 0, 12pm -> 12 (below)
    if (isPm) {
      hour += 12
    }
  } else if (hour > 23) {
    return null
  }

  return `${pad2(hour)}:${pad2(minute)}`
}

/**
 * Mirrors Rust's `str::parse::<u32>()`: ASCII digits only, an optional leading
 * `+` is allowed, a leading `-` is not, surrounding whitespace is NOT allowed,
 * an empty string fails, and anything above u32::MAX overflows to an error.
 * `Number()`/`parseInt()` are both wrong here — `Number(" 9 ")` is 9 and
 * `parseInt("9abc")` is 9, where Rust rejects both.
 */
function parseU32(s: string): number | null {
  if (!/^\+?[0-9]+$/.test(s)) {
    return null
  }
  const n = Number(s)
  if (n > U32_MAX) {
    return null
  }
  return n
}

function pad2(n: number): string {
  return n < 10 ? `0${n}` : `${n}`
}

/**
 * Mirrors Rust's `split_whitespace()`: splits on runs of whitespace and yields
 * no empty tokens (so a leading or trailing run does not produce `''`).
 */
function splitWhitespace(s: string): string[] {
  return s.split(/\s+/).filter((t) => t.length > 0)
}

// ---------------------------------------------------------------------------
// Date arithmetic — plain integer triples, never a `Date`
// ---------------------------------------------------------------------------

interface CivilDate {
  y: number
  m: number // 1-12
  d: number // 1-31
}

/** Parses a `YYYY-MM-DD` string. Throws on garbage — callers pass DB dates. */
function parseDate(s: string): CivilDate {
  const m = /^(-?\d+)-(\d{2})-(\d{2})$/.exec(s)
  if (m === null) {
    throw new Error(`recurrence: expected YYYY-MM-DD, got ${JSON.stringify(s)}`)
  }
  return { y: Number(m[1]), m: Number(m[2]), d: Number(m[3]) }
}

/**
 * Formats back to `YYYY-MM-DD`, mirroring chrono's `%Y`: 0..=9999 is
 * zero-padded to 4 digits, anything outside that range carries an explicit
 * sign. That sign is reachable through the parser — "every 100000 years" is a
 * legal rule — so the odd-looking "+102026-08-09" is what the desktop app
 * actually writes to `due_date`, and this port reproduces it rather than
 * quietly disagreeing. (Verified against chrono itself, not inferred from its
 * docs: `NaiveDate::from_ymd_opt(102026, 8, 9).format("%Y-%m-%d")` is
 * "+102026-08-09" and year -5 is "-0005-08-09".)
 */
function formatDate(date: CivilDate): string {
  const year =
    date.y >= 0 && date.y <= 9999
      ? String(date.y).padStart(4, '0')
      : date.y > 9999
        ? `+${date.y}`
        : `-${String(-date.y).padStart(4, '0')}`
  return `${year}-${pad2(date.m)}-${pad2(date.d)}`
}

/** Days since 1970-01-01, proleptic Gregorian (Howard Hinnant's days_from_civil). */
function daysFromCivil({ y, m, d }: CivilDate): number {
  const yy = y - (m <= 2 ? 1 : 0)
  const era = Math.floor(yy / 400)
  const yoe = yy - era * 400 // [0, 399]
  const doy = Math.floor((153 * (m + (m > 2 ? -3 : 9)) + 2) / 5) + d - 1 // [0, 365]
  const doe = yoe * 365 + Math.floor(yoe / 4) - Math.floor(yoe / 100) + doy // [0, 146096]
  return era * 146097 + doe - 719468
}

/** Inverse of `daysFromCivil` (civil_from_days). */
function civilFromDays(z: number): CivilDate {
  const zz = z + 719468
  const era = Math.floor(zz / 146097)
  const doe = zz - era * 146097 // [0, 146096]
  const yoe = Math.floor((doe - Math.floor(doe / 1460) + Math.floor(doe / 36524) - Math.floor(doe / 146096)) / 365)
  const y = yoe + era * 400
  const doy = doe - (365 * yoe + Math.floor(yoe / 4) - Math.floor(yoe / 100)) // [0, 365]
  const mp = Math.floor((5 * doy + 2) / 153) // [0, 11]
  const d = doy - Math.floor((153 * mp + 2) / 5) + 1 // [1, 31]
  const m = mp + (mp < 10 ? 3 : -9) // [1, 12]
  return { y: y + (m <= 2 ? 1 : 0), m, d }
}

/** Negative when `a` is before `b`, positive when after, 0 when equal. */
function compareDates(a: CivilDate, b: CivilDate): number {
  if (a.y !== b.y) return a.y - b.y
  if (a.m !== b.m) return a.m - b.m
  return a.d - b.d
}

function daysInMonth(year: number, month: number): number {
  if (month === 2) {
    const leap = (year % 4 === 0 && year % 100 !== 0) || year % 400 === 0
    return leap ? 29 : 28
  }
  return month === 4 || month === 6 || month === 9 || month === 11 ? 30 : 31
}

function addDays(date: CivilDate, days: number): CivilDate {
  return civilFromDays(daysFromCivil(date) + days)
}

/**
 * Adds N months, clamping the day-of-month to the target month's length:
 * first day of (month + N), then min(original day, days_in_that_month).
 */
function addMonthsClamped(date: CivilDate, months: number): CivilDate {
  const total = date.y * 12 + (date.m - 1) + months
  const y = Math.floor(total / 12)
  const m = total - y * 12 + 1
  return { y, m, d: Math.min(date.d, daysInMonth(y, m)) }
}

function addInterval(date: CivilDate, unit: RecurrenceUnit, interval: number): CivilDate {
  switch (unit) {
    case 'day':
      return addDays(date, interval)
    case 'week':
      return addDays(date, interval * 7)
    case 'month':
      return addMonthsClamped(date, interval)
    case 'year':
      // `interval` is already clamped to MAX_INTERVAL by the caller, so
      // interval * 12 stays well inside the safe-integer range.
      return addMonthsClamped(date, interval * 12)
  }
}

/**
 * Next due date STRICTLY after `today`, advancing by whole intervals from
 * `currentDue`. Both inputs and the result are `YYYY-MM-DD`.
 *
 * Landing exactly on `today` is not enough — it keeps advancing. Completing
 * early therefore advances one interval from the DUE date (not from the day
 * you completed it), and completing late skips every occurrence already in the
 * past in one go.
 *
 * The interval is clamped to [1, MAX_INTERVAL] here as well as in the parser,
 * so a hand-built rule with interval 0 terminates instead of looping forever.
 */
export function nextOccurrence(rule: RecurrenceRule, currentDue: string, today: string): string {
  const interval = Math.min(Math.max(Math.trunc(rule.interval), 1), MAX_INTERVAL)
  const todayDate = parseDate(today)
  let candidate = parseDate(currentDue)
  for (;;) {
    candidate = addInterval(candidate, rule.unit, interval)
    if (compareDates(candidate, todayDate) > 0) {
      return formatDate(candidate)
    }
  }
}
